yasa/plotting.py

```diff
@@ -159,23 +163,11 @@ def plot_spectrogram(
         Single-channel EEG data. Must be a 1D NumPy array.
     sf : float
         The sampling frequency of data AND the hypnogram.
-    hypno : array_like
-        Sleep stage (hypnogram), optional.
+    hyp : :py:class:`yasa.Hypnogram`
```

Comment:
This gets tricky here. This is related to your question:

> Sidenote: Currently, yasa.Hypnogram.upsample_to_data returns an array but yasa.Hypnogram.upsample returns another Hypnogram instance. Is that intended? It seems like they should both return another Hypnogram instance and then user can get the integer array as-needed. But maybe you have some other logic behind this.

- `yasa.Hypnogram.upsample`: convenient way to upsample using pandas frequency string, and return a `yasa.Hypnogram` instance. Not suited to align with EEG data.
- `yasa.Hypnogram.upsample_to_data`: Suited to align with EEG data. Returns a numpy array with integers for compatibility with most functions in YASA (spectral, events detection, etc).

My vote would be to keep `hypno` here, i.e. an upsampled numpy array with integers. However, since yasa.plot_hypnogram now requires a yasa.Hypnogram instance you will need to do an internal conversion to an instance 🤔

Unless we modify `yasa.Hypnogram.upsample_to_data` to return a class instance instead...

Reply:
By "suited to align with EEG data", I mean that there will be an automatic cropping or padding of the data to exactly match the number of samples in the EEG data

Reply:
Unrelated, even in cases where we are now expecting a yasa.Hypnogram instance, I would still keep `plot_spectrogram(..., hypno=)` and not `plot_spectrogram(..., hyp=)`. Just to limit API-breaking changes.

Reply:
Yup this isn't surprising. I figured this was worth a conversation before any changes -- I had just realized it would break unittests right before I was ready to push, and so decided to just try something and go for it. I don't have strong opinions here, and will revert to whatever you think is best.

I'll probably just revert all the `yasa.plot_spectrogram` code and then make smaller changes from there. Sounds like it would just be making a `Hypnogram` "under the hood" before plotting. lmk if you're thoughts progress on this.

Reply:
> I'll probably just revert all the yasa.plot_spectrogram code and then make smaller changes from there. Sounds like it would just be making a Hypnogram "under the hood" before plotting. lmk if you're thoughts progress on this.

Yep, let's do that for now!

Reply:
lol I shoulda started with this approach. Just a few lines and user won't notice any difference. Currently it only works for `n_stages=5` because that's the default when instantiating a Hypnogram. Maybe we add a `**hypnogram_kwargs` option in the future but I didn't for now since that use-case is probably rare and it might be sooner that this function gets reworked to just takes a Hypnogram object anyway.